src/extension/vscode/client.ts

```diff
@@ -0,0 +1,162 @@
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const clientRef = new WeakRef(client);
+  const registered = _registerClient(clientRef, url);
+  const registry = new FinalizationRegistry(registered.unregister);
+  const unregisterToken = {};
+  registry.register(client, "CLIENT_GC", unregisterToken);
+  registered.onCleanup(() => registry.unregister(unregisterToken));
+  return registered;
+}
+
+function makeErrorHandler(cleanup: (reason?: Reason) => void): EventListener {
+  return (e) => {
+    console.error("Apollo DevTools Client encountered WebSocket error:", e);
+    cleanup("WS_ERROR");
+  };
+}
+
+export function _registerClient(
+  clientRef: WeakRef<ApolloClient<any>>,
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const { signal, cleanup, registerCleanup } = getCleanupController();
+
+  function getClient() {
+    const client = clientRef.deref();
+    if (client) {
+      return getPrivateAccess(client);
+    } else {
+      cleanup("CLIENT_GC");
+    }
+  }
+
+  const ws = new WebSocket(url);
+  registerCleanup(ws.close.bind(ws, 1000));
+  ws.addEventListener("close", cleanup.bind(undefined, "WS_DISCONNECTED"), {
+    once: true,
+    signal,
+  });
+  ws.addEventListener("error", makeErrorHandler(cleanup), {
+    once: true,
+    signal,
+  });
+
+  const wsAdapter = createWsAdapter(ws, signal);
+  const wsRpcClient = createRpcClient(wsAdapter);
+  const wsRpcHandler = createRpcHandler(wsAdapter);
+  const wsActor = createActor(wsAdapter);
+  function getQueriesForClient() {
+    return getQueries(
+      getClient()?.queryManager.getObservableQueries("active") ?? new Map()
+    );
+  }
+  function getMutationsForClient() {
+    return getMutations(getClient()?.queryManager.mutationStore ?? {});
+  }
+  registerCleanup(
+    wsRpcHandler("getQueries", getQueriesForClient),
+    wsRpcHandler("getMutations", getMutationsForClient),
+    wsRpcHandler("getCache", () => getClient()?.cache.extract(true) ?? {}),
+    handleExplorerRequests(wsActor, getClient)
+  );
+
+  ws.addEventListener(
+    "open",
+    function open() {
+      const client = getClient();
+      if (!client) {
+        return;
+      }
+      wsActor.send({
+        type: "registerClient",
+        payload: {
+          id: "",
+          name:
+            "devtoolsConfig" in client ? client.devtoolsConfig.name : undefined,
+          version: client.version,
+          queryCount: getQueriesForClient().length,
+          mutationCount: getMutationsForClient().length,
+        },
+      });
+      loadErrorCodes(wsRpcClient, client.version);
+    },
+    { once: true, signal }
+  );
+
+  return {
+    connected: new Promise<void>((resolve) =>
+      ws.addEventListener("open", () => resolve(), { once: true, signal })
+    ),
+    unregister: cleanup,
+    onCleanup: registerCleanup,
```

Comment:
```suggestion
    disconnect: cleanup,
    onDisconnect: registerCleanup,
```

What about calling this `disconnect`? I think this will probably be a bit more intuitive than "cleanup" since its not immediately clear what is being cleaned up without knowing the implementation details. Thoughts?

Reply:
It's called `unregister` right now, because the function itself is named `registerClient`.

We could rename `registerClient` to `connectClientToDevTools` or something, then `disconnect` would make sense :thinking:
Not sure if I like that better, though - WDYT?

Reply:
Actually I do like that better! I had a similar thought this morning before reading your comment so I'm on board. A name like `connectClienttoDevTools` makes it immediately obvious what that block of code does without having to follow the import to understand what you're registering your client with. On top of that, it better follows with our `connectToDevTools` option naming we've had in Apollo Client for for a while so theres some symmetry there 🙂

Reply:
Then I'd suggest we make it *really* obvious with a `connectApolloClientToVSCodeDevtools`? :)

Reply:
Haha either one, but something with `connect` and `devTools` in it makes sense to me :)

Reply:
renamed to `connectApolloClientToVSCodeDevTools` and also edited the shape of the returned object